### PR TITLE
Fix RandomAnisotropy changing spatial shape of Subject with copy=False

### LINKED ISSUE #1268: Random Anisotropy transform changes the shape of the subject when copy=False
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Bug summary

Random Anisotropy transform changes the shape of the subject when copy=False.

### Code for reproduction

```python
import torchio as tio
import torch

subject = tio.Subject(
    scan=tio.ScalarImage(tensor=(torch.ones([1, 256, 256, 256])))
)

transform_copy = tio.RandomAnisotropy(
    downsampling=2.5,
    p=1,
    copy=True,
)

transform_no_copy = tio.RandomAnisotropy(
    downsampling=2.5,
    p=1,
    copy=False,
)

shape_original = subject.shape
shape_after_copy_transform = transform_copy(subject).shape
shape_after_no_copy_transform = transform_no_copy(subject).shape

print(f'{shape_original=}')
print(f'{shape_after_copy_transform=}')
print(f'{shape_after_no_copy_transform=}')
```

### Actual outcome

It reduces one of the dimensions, with the dimension changing at random, e.g.

input shape = [1, 256, 256, 256]
output shape may be [1, 103, 256, 256], [1, 256, 103, 256], or [1, 256, 256, 103]

This worked previously when I was using TorchIO: 0.19.7 and PyTorch: 2.0.1.
Error started when I upgraded to TorchIO: 0.20.4 and PyTorch: 2.6

### Error messages

```python-traceback
shape_original=(1, 256, 256, 256)
shape_after_copy_transform=(1, 256, 256, 256)
shape_after_no_copy_transform=(1, 256, 256, 103)

or 

shape_original=(1, 256, 256, 256)
shape_after_copy_transform=(1, 256, 256, 256)
shape_after_no_copy_transform=(1, 256, 103, 256)

or 

shape_original=(1, 256, 256, 256)
shape_after_copy_transform=(1, 256, 256, 256)
shape_after_no_copy_transform=(1, 103, 256, 256)
```

### Expected outcome

Shape should not change

shape_original=(1, 256, 256, 256)
shape_after_copy_transform=(1, 256, 256, 256)
shape_after_no_copy_transform=(1, 256, 256, 256)

### System info

```shell
Platform:   Windows-10-10.0.19045-SP0
TorchIO:    0.20.4
PyTorch:    2.6.0+cu126
SimpleITK:  2.4.1 (ITK 5.4)
NumPy:      2.2.2
Python:     3.12.8 | packaged by conda-forge | (main, Dec  5 2024, 14:06:27) [MSC v.1942 64 bit (AMD64)]
```

### PR DESCRIPTION
**Pull request by nicoloesch**
Fixes #1268 .

**Description**

With `copy=False`, the underlying `Image` (or multiple) will be changed in place, resulting in a change in their spatial shape and affine transform during the first `Resampling` transform in `apply_transform` of `RandomAnisotropy`. As a result, the attribute `spatial_shape` of the first `Image` in the `Subject` is changed inplace and the upsampling `Resampling` transform is not capable of restoring the original spatial shape. This results in a change in spatial shape if `copy=False` in the initialisation of the transform.

If the args and kwargs of the respective `Resampling` transforms for down- and upsampling are obtained **prior** to applying the transforms, the image will be restored appropriately irrespective of the value of `copy`.

**Note**
By changing the order of when attributes are obtained to instantiate the respective transforms, the transform itself **is not changed**. The subject/images are still downsampled to simulate the anisotropy. We are solely restoring the inplace altered images to the original resolution as intended. It therefore still follows the original intended use as outline in the [official Documentation](https://torchio.readthedocs.io/transforms/augmentation.html#randomanisotropy):

> Downsample an image along an axis and upsample to initial space.
This transform simulates an image that has been acquired using anisotropic spacing and resampled back to its original spacing.

**Checklist**

- [x] I have read the [`CONTRIBUTING`](https://github.com/TorchIO-project/torchio/blob/main/CONTRIBUTING.rst) docs and have a developer setup (especially important are `pre-commit`and `pytest`)
- [x] Non-breaking change (would not break existing functionality)
- [ ] Breaking change (would cause existing functionality to change)
- [ ] Tests added or modified to cover the changes
- [x] Integration tests passed locally by running `pytest`
- [ ] In-line docstrings updated
- [ ] Documentation updated, tested running `make html` inside the `docs/` folder
- [x] This pull request is ready to be reviewed

### CODE REVIEW

**Comment by fepegar** [comment]
Thank you, @nicoloesch!
